pkg/controller/ingress-shim/controller.go

```diff
@@ -161,7 +129,48 @@ func (c *controller) ProcessItem(ctx context.Context, key string) error {
 		return err
 	}
 
-	return c.Sync(ctx, crt)
+	return c.sync(ctx, crt)
+}
+
+// Whenever a Certificate gets deleted, we want to reconcile its parent Ingress.
+// This parent Ingress is called "controller object". For example, the following
+// Certificate is controlled by the Ingress "example":
+//
+//     kind: Certificate
+//     metadata:
+//       namespace: cert-that-was-deleted
+//       ownerReferences:
+//       - controller: true                                       ← this
+//         apiVersion: networking.k8s.io/v1beta1
+//         kind: Ingress
+//         name: example
+//         blockOwnerDeletion: true
+//         uid: 7d3897c2-ce27-4144-883a-e1b5f89bd65a
+func certificateDeleted(queue workqueue.RateLimitingInterface) func(obj interface{}) {
+	return func(obj interface{}) {
+		crt, ok := obj.(*cmapi.Certificate)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("not a Certificate object: %#v", obj))
+			return
+		}
+
+		ref := metav1.GetControllerOf(crt)
+		if ref == nil {
+			// No controller should care about orphans being deleted or
+			// updated.
+			return
+		}
+
+		// We don't check the apiVersion e.g. "networking.k8s.io/v1beta1"
+		// because there is no chance that another object called "Ingress" be
+		// the controller of a Certificate.
+		if ref.Kind != "Ingress" {
+			return
+		}
+
+		// Queue items are simple strings of the form "namespace-1/ingress-1".
+		queue.Add(crt.Namespace + "/" + ref.Name)
```

Comment:
Does the `ref` hold the namespace? If so I would use that. I realise that they should be the same, but it
just seems clearer to me that in order to process the ingress that `ref` points to we should build the
id out of `ref` alone.

Reply:
The `ref` does not hold the namespace. I added a comment to clarify this point:

```go
// Owner references don't know about the namespace of the referenced
// object. That's because owner refs do not support cross-namespace
// references. We also know that the Certificate and its parent Ingress
// must both be on the same namespace. We thus use the Certificate's
// namespace to trigger a resync of the parent Ingress (the string below
// is a "key" of the form "namespace-1/my-ingress").
queue.Add(cert.Namespace + "/" + ingress.Name)
```